File: 2953-count-complete-substrings/2953-count-complete-substrings.cpp
class Solution {
public:
    vector<vector<int>>vec;
    int func(int x,int k,string s)// Standard sliding window approach counting substrings with length x*k with each character occuring exactly k times 
	{  
        int ans=0;
        for(auto &it:vec){
            int i=it[0],j=it[0],e=it[1];
            int cntk=0;
            unordered_map<char,int>mp;
            for(j=it[0];j<=e;j++)
			{
                mp[s[j]]++;
                if(mp[s[j]]==k) cntk++;
                if(j-i+1>(x*k)){
                    mp[s[i]]--;
                    if(mp[s[i]]==(k-1)) cntk--;
                    if(mp[s[i]]==0) mp.erase(s[i]);
                    i++;
                }
                if(mp.size()==x && cntk==mp.size()) ans++;
            }
		}
        return ans;
    }
    int countCompleteSubstrings(string word, int k) {
        vector<int>v;
        string s=word;
        for(int i=1;i<s.size();i++){
            if(i<s.size() && abs(s[i]-s[i-1])>2) v.push_back(i);
        }
        
        
        
        int st=0,e=s.size()-1;
        for(auto &it:v){
            if(st<=it-1)
            vec.push_back({st,it-1});
            st=it;
        }
       if(st<=e)
       vec.push_back({st,e}); // vec will store ranges in which substring can be formed
        
       int ans=0;
       for(int i=1;i<=26 && i*k<=s.size();i++){
           ans+=func(i,k,word); // func will return number of substrings with i different characters each occuring exactly k times
       }
       return ans;
    }
};

File: 2953-count-complete-substrings/2953-count-complete-substrings.java
class Solution {
    private List<int[]> ranges;

    private int func(int x, int k, String s) {
        int ans = 0;
        for (int[] it : ranges) {
            int i = it[0], j = it[0], e = it[1];
            int cntk = 0;
            Map<Character, Integer> mp = new HashMap<>();
            for (j = it[0]; j <= e; j++) {
                mp.put(s.charAt(j), mp.getOrDefault(s.charAt(j), 0) + 1);
                if (mp.get(s.charAt(j)) == k) cntk++;
                if (j - i + 1 > (x * k)) {
                    mp.put(s.charAt(i), mp.get(s.charAt(i)) - 1);
                    if (mp.get(s.charAt(i)) == (k - 1)) cntk--;
                    if (mp.get(s.charAt(i)) == 0) mp.remove(s.charAt(i));
                    i++;
                }
                if (mp.size() == x && cntk == mp.size()) ans++;
            }
        }
        return ans;
    }

    public int countCompleteSubstrings(String word, int k) {
        ranges = new ArrayList<>();
        List<Integer> v = new ArrayList<>();
        for (int i = 1; i < word.length(); i++) {
            if (i < word.length() && Math.abs(word.charAt(i) - word.charAt(i - 1)) > 2) {
                v.add(i);
            }
        }

        int st = 0, e = word.length() - 1;
        for (int it : v) {
            if (st <= it - 1)
                ranges.add(new int[]{st, it - 1});
            st = it;
        }
        if (st <= e)
            ranges.add(new int[]{st, e});

        int ans = 0;
        for (int i = 1; i <= 26 && i * k <= word.length(); i++) {
            ans += func(i, k, word);
        }
        return ans;
    }
}